diff -up Python-2.7.5/setup.py.orig Python-2.7.5/setup.py
--- Python-2.7.5/setup.py.orig	2013-11-07 01:36:18.853604232 +0100
+++ Python-2.7.5/setup.py	2013-11-07 01:39:22.163305821 +0100
@@ -1483,6 +1483,9 @@ class PyBuildExt(build_ext):
         # Fredrik Lundh's cElementTree module.  Note that this also
         # uses expat (via the CAPI hook in pyexpat).
 
+        # Add an explicit RPATH to _elementtree.so (rhbz#1019345)
+        EXPAT_RPATH = '/usr/lib64' if sys.maxint == 0x7fffffffffffffff else '/usr/lib'
+
         if os.path.isfile(os.path.join(srcdir, 'Modules', '_elementtree.c')):
             define_macros.append(('USE_PYEXPAT_CAPI', None))
             exts.append(Extension('_elementtree',
@@ -1492,6 +1495,7 @@ class PyBuildExt(build_ext):
                                   sources = ['_elementtree.c'],
                                   depends = ['pyexpat.c'] + expat_sources +
                                       expat_depends,
+                                  extra_link_args = ['-Wl,-rpath,%s' % EXPAT_RPATH]
                                   ))
         else:
             missing.append('_elementtree')
